framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/RSPServerSocketLauncher.java

```diff
@@ -0,0 +1,120 @@
+package org.jboss.tools.rsp.server;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.net.Socket;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.eclipse.lsp4j.jsonrpc.MessageConsumer;
+import org.eclipse.lsp4j.jsonrpc.MessageProducer;
+import org.eclipse.lsp4j.jsonrpc.RemoteEndpoint;
+import org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor;
+import org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer;
+import org.jboss.tools.rsp.api.RSPClient;
+import org.jboss.tools.rsp.api.ServerSocketLauncher;
+import org.jboss.tools.rsp.server.spi.client.ClientThreadLocal;
+
+class RSPServerSocketLauncher<T> extends ServerSocketLauncher<T> {
+
+	public RSPServerSocketLauncher(Object localService, Class<T> remoteInterface, Socket socket,
+			PrintWriter tracing) throws IOException {
+		super(localService, remoteInterface, socket, tracing);
+	}
+
+	
+	/**
+	 * Start a thread that listens for messages in the message producer and forwards them to the message consumer.
+	 * 
+	 * @param messageProducer - produces messages, e.g. by reading from an input channel
+	 * @param messageConsumer - processes messages and potentially forwards them to other consumers
+	 * @param executorService - the thread is started using this service
+	 * @return a future that is resolved when the started thread is terminated, e.g. by closing a stream
+	 */
+
+	@Override
+	protected Future<Void> startListeningCalled(StreamMessageProducer messageProducer, MessageConsumer messageConsumer,
+			ExecutorService execService, Object remoteProxy, RemoteEndpoint remoteEndpoint) {
+		
+		CustomMessageProcessor reader = new CustomMessageProcessor(messageProducer, messageConsumer, remoteProxy);
+		final Future<?> result = execService.submit(reader);
+		return toFutureVoid(result, messageProducer);
+	}
+	
+	private Future<Void> toFutureVoid(Future<?> result, StreamMessageProducer messageProducer) {
+		return new Future<Void>() {
+			@Override
+			public Void get() throws InterruptedException, ExecutionException {
+				return (Void) result.get();
+			}
+
+			@Override
+			public Void get(long timeout, TimeUnit unit)
+					throws InterruptedException, ExecutionException, TimeoutException {
+				return (Void) result.get(timeout, unit);
+			}
+
+			@Override
+			public boolean isDone() {
+				return result.isDone();
+			}
+
+			@Override
+			public boolean cancel(boolean mayInterruptIfRunning) {
+				if (mayInterruptIfRunning && messageProducer instanceof Closeable) {
```

Comment:
This is simply code extracted from a super-class into a method rather than being done in-line. I have no intention on making further modifications. I only wanted to extract the logic into its own method.